File: sdks/node/src/plugins/kiroide.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import {
  AGENT_KEY,
  buildKiroIdeHook,
  kiroIdeRuntimeContracts,
  renderKiroIdeDocument,
  requireAvailableKiroIdeHooks,
  withoutManagedKiroIdeHooks,
} from './kiroide-contract.js';
import {
  commitKiroIdeInstallation,
  commitKiroIdeUninstall,
  preflightKiroIdeInstallation,
  prepareKiroIdeInstallation,
  prepareKiroIdeUninstall,
} from './kiroide-installation.js';
import { readKiroIdeSources } from './kiroide-io.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';
import { SUPPORTED_AGENTS } from './registry.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

export const kiroidePlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const sources = await readKiroIdeSources();
    requireAvailableKiroIdeHooks(sources.document.hooks);
    await preflightKiroIdeInstallation(config, sources.paths);
  },

  async install(config: InstallConfig): Promise<void> {
    const sources = await readKiroIdeSources();
    requireAvailableKiroIdeHooks(sources.document.hooks);
    const paths = await preflightKiroIdeInstallation(config, sources.paths);
    const hooks = withoutManagedKiroIdeHooks(sources.document.hooks);
    const rendered = renderKiroIdeDocument(sources.document, [
      ...hooks,
      buildKiroIdeHook('elydora-guard', paths.guardPath),
      buildKiroIdeHook('elydora-audit', paths.auditPath),
    ]);
    const prepared = await prepareKiroIdeInstallation(config, sources, rendered);
    await commitKiroIdeInstallation(prepared);
    console.log(`  Kiro IDE workspace hooks: ${sources.paths.configPath}`);
    console.log('  Kiro IDE verification: confirm Elydora hooks in the Agent Hooks panel.');
  },

  async uninstall(agentId?: string): Promise<void> {
    const sources = await readKiroIdeSources();
    const rendered = renderKiroIdeDocument(
      sources.document,
      withoutManagedKiroIdeHooks(sources.document.hooks, agentId),
    );
    const prepared = await prepareKiroIdeUninstall(sources, rendered, agentId);
    await commitKiroIdeUninstall(prepared);
  },

  async status(): Promise<PluginStatus> {
    const sources = await readKiroIdeSources();
    requireAvailableKiroIdeHooks(sources.document.hooks);
    const contracts = kiroIdeRuntimeContracts(sources.document.hooks);
    const hookConfigured = contracts.length > 0;
    const hookScriptExists = hookConfigured
      && (await Promise.all(contracts.map((contract) => (
        managedRuntimeFilesExist(contract, AGENT_KEY)
      )))).every(Boolean);
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: sources.paths.configPath,
    };
  },
};

File: sdks/go/cmd/elydora/plugins/kiroide.go
package plugins

import "fmt"

// KiroIdePlugin manages Kiro IDE 1.0 workspace hooks.
type KiroIdePlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Kiro IDE commits both generated runtimes,
// runtime metadata, credentials, workspace hooks, and legacy migration in one
// transaction.
func (p *KiroIdePlugin) ManagesGuardRuntime() bool {
	return true
}

// PreflightInstall validates every source and runtime identity before writes.
func (p *KiroIdePlugin) PreflightInstall(config InstallConfig) error {
	sources, err := readKiroIdeSources()
	if err != nil {
		return err
	}
	if err := requireAvailableKiroIdeHooks(sources.document.hooks); err != nil {
		return err
	}
	_, _, err = preflightKiroIdeInstallation(config)
	return err
}

func (p *KiroIdePlugin) Install(config InstallConfig) error {
	sources, err := readKiroIdeSources()
	if err != nil {
		return err
	}
	if err := requireAvailableKiroIdeHooks(sources.document.hooks); err != nil {
		return err
	}
	paths, nodePath, err := preflightKiroIdeInstallation(config)
	if err != nil {
		return err
	}
	hooks := withoutManagedKiroIdeHooks(sources.document.hooks, "")
	guard, err := buildKiroIdeHook(kiroIdeGuardName, nodePath, paths.guardPath)
	if err != nil {
		return err
	}
	audit, err := buildKiroIdeHook(kiroIdeAuditName, nodePath, paths.auditPath)
	if err != nil {
		return err
	}
	rendered, err := renderKiroIdeDocument(
		sources.document,
		append(hooks, guard, audit),
	)
	if err != nil {
		return err
	}
	prepared, err := prepareKiroIdeInstallation(config, sources, paths, rendered)
	if err != nil {
		return err
	}
	if err := commitKiroIdeInstallation(prepared, p.rename); err != nil {
		return err
	}
	fmt.Printf("Kiro IDE workspace hooks: %s\n", sources.paths.configPath)
	fmt.Println("Kiro IDE verification: confirm both Elydora entries in the Agent Hooks panel.")
	return nil
}

func (p *KiroIdePlugin) Uninstall(agentID string) error {
	sources, err := readKiroIdeSources()
	if err != nil {
		return err
	}
	rendered, err := renderKiroIdeDocument(
		sources.document,
		withoutManagedKiroIdeHooks(sources.document.hooks, agentID),
	)
	if err != nil {
		return err
	}
	prepared, err := prepareKiroIdeUninstall(sources, rendered, agentID)
	if err != nil {
		return err
	}
	if err := commitKiroIdeUninstall(prepared, p.rename); err != nil {
		return err
	}
	fmt.Println("Uninstalled Elydora hooks from Kiro IDE.")
	return nil
}

func (p *KiroIdePlugin) Status() (PluginStatus, error) {
	sources, err := readKiroIdeSources()
	entry := SupportedAgents[kiroIdeAgentKey]
	status := PluginStatus{
		AgentName:   kiroIdeAgentKey,
		DisplayName: entry.Name,
		ConfigPath:  sourcesPathOrEmpty(sources),
	}
	if err != nil {
		return status, err
	}
	if err := requireAvailableKiroIdeHooks(sources.document.hooks); err != nil {
		return status, err
	}
	contracts, err := kiroIdeRuntimeContracts(sources.document.hooks)
	if err != nil {
		return status, err
	}
	status.HookConfigured = len(contracts) > 0
	if !status.HookConfigured {
		return status, nil
	}
	status.HookScriptExists, err = kiroIdeRuntimeFilesExist(contracts)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}

func sourcesPathOrEmpty(sources *kiroIdeSources) string {
	if sources == nil {
		return ""
	}
	return sources.paths.configPath
}
